### DISCUSSION TITLE
외부 인터넷이 한정적으로 연결되는 환경에서 "다이나믹 서브셋" 적용하는 방법 문의

### DISCUSSION THREAD

**Discussion by piarey** [orioncactus/pretendard#22] [Q&A] [answered]
안녕하세요. 프리덴다드 체 잘 사용중입니다. 감사합니다.

**[질문]**

1.컴퓨터 fonts 폴더에 설치할 수 있는 설치파일은 별도로 없는 것일까요?
src: local('Pretendard Black') 
요걸 적용하려면 개인 컴퓨터에 설치가 되어야 하는데 말이죠.
**개인 컴퓨터에 설치를 어떤식으로 하면 될까요?**

<br><br>
====================
<br><br>

2.다이나믹 서브셋 CDN 제공해주신 코드를 보면
src: local('Pretendard Black') 
와 같이 사용자 개인 컴퓨터에 설치되어있다면 그 폰트를 적용! 이라는 코드가 없는데요.

=> 
**다이나믹 서브셋 CDN 은 local 은 따로 지원은 안해주시고
무조건 인터넷 연결이 된 환경에서 CDN 설치방식으로밖에 안될까요?**

<br><br>
====================
<br><br>

3.다이나믹 서브셋을 CDN 으로 하는게 아니라
**회사 서버에 올리려면 어떤 방식으로 해야할까요???**

=>
[Pretendard-1.1.1\web\static\woff2-subset]
폴더 들어가보면 "Pretendard-Black.subset.woff2" 형식으로만 되어있는데요~

이게 이렇게가 아니라
woff2-dynamic-subset/Pretendard-Thin.subset.0.woff2
woff2-dynamic-subset/Pretendard-Thin.subset.1.woff2
woff2-dynamic-subset/Pretendard-Thin.subset.2.woff2
woff2-dynamic-subset/Pretendard-Thin.subset.3.woff2
woff2-dynamic-subset/Pretendard-Thin.subset.4.woff2
....
이런식으로 woff2 파일들이 다 분리되어있어야 하지 않나요?

<br><br>
====================
<br><br>

4.내부 프로젝트에 "다이나믹 서브셋 - HTML" 방식으로 폰트 적용을 하려고 하였는데요.
이게 보안 상 내부망에서는 인터넷 연결이 한정적으로 되어있어서
CDN 으로는 적용이 안되더라고요.

=> 
방법 A :
CDN 방식으로 넣고,
내부망 컴퓨터에 폰트파일을 설치한다.
어차피 "src: local('Pretendard Black')" 로 폰트가 내부컴퓨터에 설치되어있다면 적용 될것 같다.
(단, @font-face 에 "src: local('Pretendard Black')" 이 적용되어있어야하는데, 다이나믹 서브셋에는 없다....)

방법 B : 
프리텐다드 폰트를 제가 다운받아서
회사 서버에 폰트파일 올리는 식으로 하면 될것 같은데요.
요것도.... 4번 질문이랑 연결되는데, 다이나믹 서브셋을 어떤방식으로 올릴지 의문입니다.

방법 C : 
기타 더 좋은 방법 있으면 추천 부탁드립니다.

감사합니다.

**Comment by piarey** [comment]
+ [추가질문]

1.[Pretendard-1.1.1\web\static\woff2-subset]
폴더 내에 있는 Pretendard-Black.subset.woff 을 
고대로 font-face 로 불러와서 적용하면,,,, 다이나믹 서브셋 폰트가 적용되는건가요?

2.subset 파일 쪼개놓은것들은 제가 지원 받을 수 있을까요?

**Comment by orioncactus** [comment, marked as answer]
안녕하세요! Pretendard에 관심을 가져주셔서 감사드립니다.

1. 로컬에서 Pretendard를 사용하시려면 dist/public/static에 있는 글꼴을 설치하시면 됩니다.

2. Pretendard의 동적 서브셋은 Google Fonts에 제공하는 Noto Sans KR의 동적 서브셋과 동일한 구조로 가져왔기 때문에, 말씀하신 local 속성이 빠져 있습니다. 다만 비슷하게 동적 서브셋을 제공하는 IBM Plex KR의 경우를 확인해보니 local 속성이 포함되어 있어 해당 속성이 글꼴 표시 성능 향상에 도움이 되는지 확인해본 뒤 Pretendard에 적용하려고 합니다. 

3. 가변 글꼴을 사용해야 하는 상황이 아니라면 dist/web/static에 있는 파일을 그대로 회사 서버에 올리면 됩니다. 회사에서 Pretendard의 표시 용량을 줄이고자 한다면 동일하게 pretendard-dynamic-subset.css를 웹에 적용해 사용하실 수 있습니다.

4. font-face에서 local 속성은 font-face를 불러오는 환경에서 Pretendard가 이미 설치되어 있으면 서버에서 가져오지 않고 내부 글꼴로 표시하는 속성이기 때문에, 내부 서버에서 otf 또는 ttf를 올려둔다고 해서 그 파일로 표시되거나 하지는 않습니다. 따라서 올바른 방법은, 동적 서브셋과 같이 분할된 otf 파일을 로컬에 적용할 수 있게 제공하는 것이 아닌, 3번과 같은 형태로 서버에 올려 웹에 적용하는 것이며, 혹시 이러한 웹폰트를 서버에 적용하는 부분에 이해가 어려우시다면 다시 알려드리겠습니다.

추가적으로 질문 주신 것처럼 동적 서브셋만 사용하시고자 한다면 dist/web/static 디렉터리에서 pretendard-dynamic-subset.css와 woff2-dynamic-subset(woff-dynamic-subset) 폴더를 서버에 올린 뒤 사용하시면 됩니다.

질문에 도움이 되었길 바랍니다. 감사합니다.

**Comment by orioncactus** [comment]
이 이슈는 이제 종결 처리하고 Discussion으로 옮기도록 하겠습니다. 또 다른 질문이 있으시면 언제든 알려주세요!

**Reply by pubaooooo** [reply]
안녕하세요. Pretendard의 dynamic-subset 폰트를 사용하고자하여서 해당 문의글을 참고하였는데
한가지 궁금한 점이 있어서 댓글남깁니다. 
dist/web/static 디렉터리에서 woff2-dynamic-subset(woff-dynamic-subset) 폴더는 어디있는건가요?

**Reply by orioncactus** [reply]
다이나믹 서브셋은 레포지터리를 클론하거나 레포지터리 zip을 다운로드해 사용하셔야 합니다. 일반 사용자가 사용하는 zip 파일에는 포함해두지 않았습니다.

**Comment by clichedmoog** [comment]
안녕하세요. 여기 이렇게 discussions에 해당하는 내용이 있었군요!
저도 필요해서 repo 를 살펴보다가 로컬에 이미 설치된 폰트를 우선 사용하도록 하는 방안에 대해 의견을 남깁니다.

@import로 웹 폰트를 로드하고 있는데, 사용자 시스템에 Pretendard가 이미 설치되어 있는 경우에도 네트워크를 통해 폰트를 다시 다운로드하게 됩니다.

성능 최적화 관점에서 local() 지시자를 활용하는 것은 어떨까요?

```css
@font-face {
  font-family: 'Pretendard';
  src: local('Pretendard'), url(...기존 코드...);
}
```

이렇게 하면:
- 로컬에 설치된 폰트가 있으면 우선 사용 (네트워크 요청 없음)
- 없는 경우에만 웹 폰트를 다운로드
- 초기 로딩 성능 개선 효과

전문적인 의견은 아니기에 검토 부탁드려요!
